myapp/views.py

```diff
@@ -1,51 +1,85 @@
+from django.contrib.auth import logout
+from .exceptions.stock_service import StockServerUnReachable, StockSymbolNotFound
+
+
+# View for the home page - a list of 20 of the most active stocks
+def index(request):
+    # Query the stock table, filter for top ranked stocks and order by their rank.
+    data = Stock.objects.filter(top_rank__isnull=False).order_by('top_rank')
+    return render(request, 'index.html', {'page_title': 'Main', 'data': data})
+
+
+# View for the single stock page
+# symbol is the requested stock's symbol ('AAPL' for Apple)
+def single_stock(request, symbol):
+    context = {}
+    status_code = 200
+    template = 'single_stock.html'
+    try:
+        data = stock_api.get_stock_info(symbol)
+    except StockSymbolNotFound as e:
+        status_code = 404  # stock symbol not found!
+        context = {'error_message': e.message, "status_code": status_code}
+        template = "exception.html"
+    except StockServerUnReachable as e:
+        status_code = 503  # Service Unavailable code
+        context = {'error_message': e.message, "status_code": status_code}
+        template = "exception.html"
+    except Exception as e:
+        status_code = 520  # Unknown Error
+        context = {'error_message': "Unknown Error occurred: {}".format(", ".join(e.args)), "status_code": status_code}
+        template = "exception.html"
+    else:
+        context = {'page_title': 'Stock Page - %s' % symbol, 'data': data}
+    finally:
+        resp = render(request, template, context)
+        resp.status_code = status_code
+        return resp
+
+
+def register(request):
+    # If post -> register the user and redirect to main page
+    if request.method == 'POST':
+        firstname = request.POST.get('firstname')
+        lastname = request.POST.get('lastname')
+        email = request.POST.get('email')
+        password = request.POST.get('password')
+
+        newuser = User.objects.create_user(username=email, email=email, password=password)
+        newuser.first_name = firstname
+        newuser.last_name = lastname
+        newuser.save()
+        return redirect('index')
+    else:
+        # If not post (regular request) -> render register page
+        return render(request, 'register.html', {'page_title': 'Register'})
+
+
+def logout_view(request):
+    logout(request)
+    return redirect('index')
+
+
+# API for a stock's price over time
+# symbol is the requested stock's symbol ('AAPL' for Apple)
+# The response is JSON data of an array composed of "snapshot" objects (date + stock info + ...), usually one per day
+def single_stock_historic(request, symbol):
+    data = stock_api.get_stock_historic_prices(symbol, time_range='1m')
+    return JsonResponse({'data': data})
+
+
+def page_not_found(request):
+    # checking if the request url path doesn't trailing with / character
+    # if not then redirecting to the same url with / trailer
+    if request.path[-1] != '/':
+        return redirect(request.path + '/')
```

Comment:
i noticed also this causing a bug here, the "/admin/" page would be not accessible, because I added the URL pattern of "page_not_found" inside the <app>(myapp)/urls, and not inside the <project>(myrails)/urls

Reply:
I think it is currently fine if the admin page is accessible since we are still building the website and it hasn't gone live